docs(contributing): update contributing guidelines (#91)
autowarefoundation/autoware-documentation: docs/contributing/index.md

To see the schedule for upcoming working group meetings, refer to the [Autoware Foundation events calendar](https://calendar.google.com/calendar/u/0/embed?src=autoware.org_6lol0ho5ft0217h8c60pi1fm30@group.calendar.google.com).

### Bug reports

Before you report a bug, please search the issue tracker for the appropriate repository. It is possible that someone has already reported the same issue and that workarounds exist. If you can't determine the appropriate repository, ask the maintainers for help by creating a new discussion in the [Q&A category](https://github.com/autowarefoundation/autoware/discussions/new?category=q-a).

When reporting a bug, you should provide a minimal set of instructions to reproduce the issue. Doing so allows us to quickly confirm and focus on the right problem.

If you want to fix the bug by yourself that will be appreciated, but you should discuss possible approaches with the maintainers in the issue before submitting a pull request.

> [Creating an issue is straightforward])<https://docs.github.com/en/issues/tracking-your-work-with-issues/creating-an-issue#creating-an-issue-from-a-repository>), but if you happen to experience any problems then create a Q&A discussion to ask for help.

Reviewer: Sorry for making comments after merged. This link is broken.

Reviewer: Opened https://github.com/autowarefoundation/autoware-documentation/pull/97

As a contributor, here are the guidelines that we would like you to follow for Autoware and its associated repositories, all of which are hosted in the [Autoware Foundation Organization](https://github.com/autowarefoundation) on GitHub.

- [Code of Conduct](#code-of-conduct)
- [What Should I Know Before I Get Started?](#what-should-i-know-before-i-get-started)
  - [Autoware Concepts](#autoware-concepts)
  - [Contributing to Open Source Projects](#contributing-to-open-source-projects)
- [How Can I Get Help?](#how-can-I-get-help)
- [How Can I Contribute?](#how-can-i-contribute)
  - [Participate in Discussions](#discussions)
  - [Join a Working Group](#working-groups)
  - [Report Bugs](#reporting-bugs)
  - [Make a Pull Request](#pull-requests)

Like Autoware itself, these guidelines are being actively developed and suggestions for improvement are always welcome! Guideline changes can be proposed by [creating a discussion in the Ideas category](https://github.com/autowarefoundation/autoware/discussions/new?category=ideas).

## Code of Conduct

> To ensure our community stays open and inclusive, our [code of conduct](https://github.com/autowarefoundation/autoware/blob/main/CODE_OF_CONDUCT.md) adheres to the [Contributor Covenant](https://www.contributor-covenant.org/) that has been adopted by [many other open source communities](https://www.contributor-covenant.org/adopters/). By participating in the Autoware community, you agree to follow this code.

Reviewer: I felt that adding three links in one line is redundant. The third one might be removed if not really necessary.

Author: All three links were present in the original contributing guide so I will leave this one for @kenji-miyake to comment on.

Author: To present my own opinion, I think that it is fine to remove the third link. Since the code of conduct itself links to the Contributor Covenant, we could simplify the section further:

```suggestion
To ensure our community stays open and inclusive, please follow our [code of conduct](https://github.com/autowarefoundation/autoware/blob/main/CODE_OF_CONDUCT.md).
```

Author: Reading the Google Developers style guide, `please` can be used when asking the reader to do something that "benefits us, inconveniences a reader, or suggests a potential issue with a product."

Asking a contributor to read the (relatively long) code of conduct could be considered an inconvenience, so I think it is okay to use `please` in this circumstance.

Reviewer 2: I see, thank you! :+1:

### Working groups

The [various working groups](https://github.com/autowarefoundation/autoware-projects/wiki#working-group-list) within the Autoware Foundation are responsible for managing low level activities as directed by the foundation's Technical Steering Committee. These working groups are open to everyone, and joining a particular working group will allow you to gain an understanding of current projects, see how those projects are managed within each group and contribute to issues that will help progress a particular project.

To see the schedule for upcoming working group meetings, refer to the [Autoware Foundation events calendar](https://calendar.google.com/calendar/u/0/embed?src=autoware.org_6lol0ho5ft0217h8c60pi1fm30@group.calendar.google.com).

### Reporting bugs

Before you submit an issue, please search the issue tracker for the appropriate repository. It is possible that someone has already reported the same issue and that workarounds exist. If you can't determine the appropriate repository, ask the maintainers for help by creating a new discussion in the [Q&A category](https://github.com/autowarefoundation/autoware/discussions/categories/q-a).

> When reporting a bug, you should provide a minimal set of instructions to reproduce the issue. It allows us to quickly confirm and focus on the right problem.```

Author: ```suggestion
When reporting a bug, you should provide a minimal set of instructions to reproduce the issue. Doing so allows us to quickly confirm and focus on the right problem.```
```

On review, I realised that the start of the second sentence should have been slightly different.

> When reporting a bug, you should provide a minimal set of instructions to reproduce the issue. Doing so allows us to quickly confirm and focus on the right problem.```

Author: ```suggestion
When reporting a bug, you should provide a minimal set of instructions to reproduce the issue. Doing so allows us to quickly confirm and focus on the right problem.
```

> In order to fix a bug, we need to reproduce and confirm it first and so you should provide a minimal reproduction. Having a minimal reproducible scenario allows us to quickly confirm a bug (or point out a coding or configuration problem) as well as confirm that we are fixing the right problem.

Reviewer: I think this comes from another page. 
How about making it shorter?

```
When reporting a bug, you should provide instructions to reproduce it. It allows us to quickly focus on the right problem.
```

Author: I think the word `minimal` should be retained, because the original reproduction scenario could be a complex environment with multiple ECUs that cannot be easily set up by the maintainers. How about this?

```suggestion
When reporting a bug, you should provide a minimal set of instructions to reproduce the issue. It allows us to quickly confirm and focus on the right problem.```

Reviewer: Looks so good!

> If you want to fix the bug by yourself, then that will be appreciated, but you should discuss possible approaches with the maintainers first before submitting a pull request.

Reviewer: I think this can be simplified as well. It would be even better if it has where contributors should make a discussion. (Q&A in GitHub discussions?) @kenji-miyake 

```
We appreciate if you fix a bug by yourself. Please discuss possible approaches with maintainers before sending a pull request
```

Author: @IshitaTakeshi @kenji-miyake My assumption was that the approach would be discussed in the issue itself, since this is where people will report confirmed and reproduceable bugs to the maintainers.

Author: How about this?

```suggestion
Providing a fix for your reported bug will be highly appreciated, but discuss possible approaches with the maintainers in the issue before submitting a pull request.
```

Reviewer: LGTM
@kenji-miyake How about this?

Reviewer 2: Considering this guideline, I prefer the original sentence because people want to fix a bug here.
https://developers.google.com/style/clause-order

Author: Done!

> You can file new issues by selecting from our new issue templates and filling out the issue template.

Reviewer: @kenji-miyake Do you mean a different verb than `file`?

Author: @IshitaTakeshi This was taken from the angular contributing guide, but it is a valid use of the verb. However, if it is confusing then it can be replaced with `create`.

Reviewer: @LalithVipulananthan OK. Thank you for teaching

Reviewer 2: @LalithVipulananthan I feel `create` is easier to understand for most people (at least for Japanese).

https://developers.google.com/style/translation#clear-language

Reviewer 2: But I feel the relationship to the previous sentence seems a bit unclear.
I believe what we want to say here is `create an issue and discuss there`, but this sentence doesn't contain the word `discuss`.

Author: Would it be better if I replaced this sentence with a short explanation of the process to create a new issue? 

It is assumed that people know what to do, but it isn't actually written anywhere (click on the Issues tab for the repository, click "New issue" then select the appropriate template)

Reviewer 2: > Would it be better if I replaced this sentence with a short explanation of the process to create a new issue?

Yes, it's nice. I think we don't need so detailed steps. Just saying "create an issue" and "if you need help, create a discussion thread in the Q&A category (with the link)" is enough for me.

Author: Updated this part to say `create` rather than `file`, linked to the GitHub documentation on how to create an issue and asked contributors to create a Q&A discussion if any problems are experienced.

### Pull requests

Pull requests can be submitted directly for small changes such as:

- Minor documentation updates.
- Fixing spelling mistakes.
- Fixing CI failures.
- Fixing warnings detected by compilers or analysis tools.
- Making small changes to a single package.

> If your pull request is not covered by the examples above or is a large change, create an issue or a discussion thread to discuss possible approaches with the maintainers before submitting a pull request. Doing so ensures that the proposed change will be in line with Autoware's design philosophy and current development plans. Lastly, ensure that you use the appropriate pull request template for the target repository.

Reviewer: I think it is better to include where to discuss.

Author: @kenji-miyake Should people use the the Q&A category to ask the maintainers about large changes, or changes that are not covered by the list of small changes?

Author: I think we should recommend that people create a Q&A discussion for large changes, so that issues are kept for confirmed bugs only.

```suggestion
If your pull request is a large change not covered by the examples above, create a Q&A discussion thread to confirm with the maintainers before submission. Doing so ensures that the proposed change is in line with Autoware's design philosophy and current development plans. Lastly, ensure that you use the appropriate pull request template for the target repository.
```

Reviewer 2: > @kenji-miyake Should people use the the Q&A category to ask the maintainers about large changes, or changes that are not covered by the list of small changes?

@LalithVipulananthan For newbies, I think yes. Or sometimes they can also use `Design`.
For accustomed contributors, we can use Discord or Issues.

So I think it's good to write like "if you're not sure where to discuss, ask in the Q&A category".

# Contributing

> First off, thank you for your interest in contributing to Autoware! Like any other open-source project, Autoware can only continue to thrive through the work of people like yourself, and all types of contribution are welcome no matter how big or small.

Author: Would something like this be okay?

```suggestion
Thank you for your interest in contributing! Autoware is supported by people like you, and all types and sizes of contribution are welcome.
```

> First off, thank you for your interest in contributing to Autoware! Like any other open-source project, Autoware can only continue to thrive through the work of people like yourself, and all types of contribution are welcome no matter how big or small.

Reviewer 2: Yeah, LGTM.

> As a contributor, here are the guidelines that we would like you to follow for Autoware and its associated repositories, all of which are hosted in the [Autoware Foundation Organization](https://github.com/autowarefoundation) on GitHub.

Author: Updated!

- [Code of Conduct](#code-of-conduct)
- [What Should I Know Before I Get Started?](#what-should-i-know-before-i-get-started)
  - [Autoware Concepts](#autoware-concepts)
  - [Contributing to Open Source Projects](#contributing-to-open-source-projects)
- [How Can I Get Help?](#how-can-I-get-help)
- [How Can I Contribute?](#how-can-i-contribute)
  - [Participate in Discussions](#discussions)
  - [Join a Working Group](#working-groups)
  - [Report Bugs](#reporting-bugs)

>   - [Make a Pull Request](#pull-requests)

Author: Good point. I think this was copied from Atom and Angular's guides. I will change this to sentence case.

  - [Participate in Discussions](#discussions)
  - [Join a Working Group](#working-groups)
  - [Report Bugs](#reporting-bugs)
  - [Make a Pull Request](#pull-requests)

- [Coding guidelines](coding-guidelines/index.md): for contributing source code
- [Testing guidelines](testing-guidelines/index.md): for contributing tests
- [Documentation guidelines](documentation-guidelines/index.md): for contributing documents
- [Pull request guidelines](pull-request-guidelines/index.md): for submitting pull requests
- [Discussion guidelines](discussion-guidelines/index.md): for discussing with other contributors
- [Support guidelines](../help/support-guidelines.md): for finding the right channel when you need help
Like Autoware itself, these guidelines are being actively developed and suggestions for improvement are always welcome! Guideline changes can be proposed by [creating a discussion in the Ideas category](https://github.com/autowarefoundation/autoware/discussions/new?category=ideas).

> It is recommended to read the [Contributing pages in ROS 2 Docs](https://docs.ros.org/en/rolling/Contributing.html) as well because Autoware's guidelines generally depend on them.

Author: I apologise for not explaining the removal - I got mixed up and thought that I was making that change with my updates for the Support pages that is still in progress. 

So, the ROS 2 docs page has a lot of information that is specific to ROS 2, as well as the information about being respectful of prior work etc, but we are asking Autoware contributors to read all of the page anyway. 

If there is something that we want to include in our own contribution guide, then I think that it would be better to explicitly add that content instead of the link.

> It is recommended to read the [Contributing pages in ROS 2 Docs](https://docs.ros.org/en/rolling/Contributing.html) as well because Autoware's guidelines generally depend on them.

Reviewer 2: Thank you, agreed!

> ### Autoware concepts

Author: I can see that, but at the same time, the introduction gives a brief explanation of each section in the documentation and so including a link to that page feels like we are telling contributors that they have to read the entire site. 🤔 

I am in the process of editing the file to include your other comments, but please allow me some more time to think about this section.

> ### Autoware concepts

Reviewer 2: No problem, thank you!
There is no need to fix all in this PR.

- [Autoware architecture](https://autowarefoundation.github.io/autoware-documentation/main/design/)
- [Autoware concepts](https://autowarefoundation.github.io/autoware-documentation/main/design/autoware-concepts/)

For experienced developers, the [Autoware interfaces](https://autowarefoundation.github.io/autoware-documentation/main/design/autoware-interfaces/) and [individual component pages](https://autowarefoundation.github.io/autoware-documentation/main/design/autoware-interfaces/components/) should also be reviewed to understand the inputs and outputs for each component or module at a more detailed level.

### Contributing to open source projects

> If you are new to open source projects, we recommend reading GitHub's [How to Contribute to Open Source guide](https://opensource.guide/how-to-contribute/) for an overview of why people contribute to open source projects, what it means to contribute and much more besides.

Reviewer 2: If possible, I'd like to add https://docs.ros.org/en/rolling/Contributing.html here.

Reviewer 2: https://github.com/autowarefoundation/autoware-documentation/pull/91#discussion_r859272117

Author: This goes back to the reason why I removed the ROS 2 docs link. Ideally we should include the information that we think is relevant in our contributing guidelines, but at the very least we should indicate which specific information on that page should be reviewed by contributors? 

I'm guessing it's the Tenets section? 🤔

Reviewer 2: Hmm, probably yes. I love the Tenets section!

Author: Okay. I think this might also have to be part of a follow-up PR. 

There are a lot of good points in the ROS 2 contribution tenets that can be included in our (😅 ) own contribution guidelines. However, I think the wording will need to be adjusted in a few places. 

For example, Open Robotics is described as being a gate-keeper. Although I understand the intent behind that description, the word `gate-keeper` has negative connotations (something that evokes a feeling or idea beyond the literal/primary meaning) so I would prefer to avoid using it.

Reviewer 2: Thank you! I'll leave it to you. :pray:

## How can I get help?

> Do not open issues for general support questions as we want to keep GitHub issues for confirmed bug reports. Instead, open a discussion in the Q&A category. For more details on our support mechanisms and links to troubleshooting guides, refer to our [Support guidelines](../support/index.md).

Reviewer 2: our -> the

But `the support mechanisms` sounds a bit unclear? :thinking: 
Should we use `Autoware's` or something like that?

```suggestion
Do not open issues for general support questions as we want to keep GitHub issues for confirmed bug reports. Instead, open a discussion in the Q&A category. For more details on the support mechanisms and links to troubleshooting guides, refer to the [Support guidelines](../support/index.md).
```

Author: Is this better?

```suggestion
Do not open issues for general support questions as we want to keep GitHub issues for confirmed bug reports. Instead, open a discussion in the Q&A category. For more details on the support mechanisms for Autoware, refer to our [Support guidelines](../support/index.md).
```

Reviewer 2: Yeah, it's nice. Are there any reasons that you used `our` for `our [Support guidelines]`?

Author: Oops. I updated the suggestion!

Reviewer 2: Perfect!

!!! note

    Issues created for questions or unconfirmed bugs will be moved to GitHub discussions by the maintainers.

## How can I contribute?

### Discussions

You can contribute to Autoware by facilitating and participating in discussions, such as:

- [Proposing a new feature to enhance Autoware](https://github.com/orgs/autowarefoundation/discussions/categories/feature-requests).
- [Joining an existing discussion and expressing your opinion](https://github.com/orgs/autowarefoundation/discussions).
- Organizing discussions for other contributors.
- [Answering questions and supporting other contributors](https://github.com/autowarefoundation/autoware/discussions/categories/q-a?discussions_q=category%3AQ%26A+is%3Aunanswered).

### Working groups

> The [various working groups](https://github.com/autowarefoundation/autoware-projects/wiki#working-group-list) within the Autoware Foundation are responsible for managing low level activities as directed by the foundation's Technical Steering Committee. These working groups are open to everyone, and joining a particular working group will allow you to gain an understanding of current projects, see how those projects are managed within each group and contribute to issues that will help progress a particular project.

Reviewer 2: I'm not sure whether it's appropriate to use the word `low level activities` here.
I feel we can rephrase it. (But I'm sorry I don't have a good idea.)

Author: Maybe `task-based activities`? 🤔

Reviewer 2: Hmm, I'd like to take care not to feel people that "TSC decide the direction and WGs simply do it", in order to motivate people. :thinking: 

So, how about `accomplishing missions given by the TSC` or something like that?

Author: That seems fine to me, but I think `goals` would be better than `missions`. If we used `goals` instead, would that still be okay?

Reviewer 2: Sure!

Author: Revised suggestion:

```suggestion
The [various working groups](https://github.com/autowarefoundation/autoware-projects/wiki#working-group-list) within the Autoware Foundation are responsible for accomplishing goals set by the Technical Steering Committee. These working groups are open to everyone, and joining a particular working group will allow you to gain an understanding of current projects, see how those projects are managed within each group and to contribute to issues that will help progress a particular project.
```

To see the schedule for upcoming working group meetings, refer to the [Autoware Foundation events calendar](https://calendar.google.com/calendar/u/0/embed?src=autoware.org_6lol0ho5ft0217h8c60pi1fm30@group.calendar.google.com).

> ### Reporting bugs

Reviewer 2: To be consistent with other headers.
Or you can change other headers like `Joining working groups`.

```suggestion
### Bug reports
```

Before you submit an issue, please search the issue tracker for the appropriate repository. It is possible that someone has already reported the same issue and that workarounds exist. If you can't determine the appropriate repository, ask the maintainers for help by creating a new discussion in the [Q&A category](https://github.com/autowarefoundation/autoware/discussions/categories/q-a).

When reporting a bug, you should provide a minimal set of instructions to reproduce the issue. Doing so allows us to quickly confirm and focus on the right problem.

If you want to fix the bug by yourself, then that will be appreciated, but you should discuss possible approaches with the maintainers first before submitting a pull request.

You can file new issues by selecting from our new issue templates and filling out the issue template.

### Pull requests

> Pull requests can be submitted directly for small changes such as:

Reviewer 2: Are there any reasons you used passive voice here?
https://developers.google.com/style/voice?hl=en

Author: Good point. I have changed this back to active voice.

- Minor documentation updates.
- Fixing spelling mistakes.
- Fixing CI failures.
- Fixing warnings detected by compilers or analysis tools.
- Making small changes to a single package.

If your pull request is not covered by the examples above or is a large change, create an issue or a discussion thread to discuss possible approaches with the maintainers before submitting a pull request. Doing so ensures that the proposed change will be in line with Autoware's design philosophy and current development plans. Lastly, ensure that you use the appropriate pull request template for the target repository.

We have some additional guidelines specific to the type of change being made, so please make sure to review those before submitting your pull request.

- [Coding guidelines](coding-guidelines/index.md)
- [Testing guidelines](testing-guidelines/index.md)

> - [Documentation guidelines](documentation-guidelines/index.md)

Reviewer 2: These items duplicate the number 3 [here](https://autowarefoundation.github.io/autoware-documentation/pr-91/contributing/pull-request-guidelines/#general-pull-request-workflow).

Do you want to write them here as well?

Author: Ah. Good point. In that case I will remove these and just leave the link to the PR guidelines.

Do not open issues for general support questions as we want to keep GitHub issues for confirmed bug reports. Instead, open a discussion in the Q&A category. For more details on our support mechanisms and links to troubleshooting guides, refer to our [Support guidelines](../support/index.md).

!!! note

> Issues created for questions or unconfirmed bugs will be moved to GitHub discussions by the maintainers.

Reviewer 2: @LalithVipulananthan You should add an empty line here.
![image](https://user-images.githubusercontent.com/31987104/165219998-50851ad9-54cf-4787-9f6d-bca73e173549.png)

Reviewer 2: Applied in c809daeef376dc94c28bd96437d2415a1a74db5e
